arrow2_convert/src/deserialize.rs

```diff
@@ -48,23 +48,36 @@ pub trait ArrowDeserialize: ArrowField + Sized {
     /// something like for<'a> &'a T::ArrayType: IntoIterator<Item=Option<E>>,
     /// However, the E parameter seems to confuse the borrow checker if it's a reference.
     fn arrow_deserialize_internal(
-        v: <<Self::ArrayType as IterRef>::Iter<'_> as Iterator>::Item,
+        v: <<Self::ArrayType as RefIntoIterator>::Iterator<'_> as Iterator>::Item,
     ) -> <Self as ArrowField>::Type {
         Self::arrow_deserialize(v).unwrap()
     }
-}
-
-/// Internal trait used to support deserialization and iteration of structs, and nested struct lists
-///
-/// Trivial pass-thru implementations are provided for arrow2 arrays that auto-implement IterRef.
-///
-/// The derive macro generates implementations for typed struct arrays.
-#[doc(hidden)]
-pub trait ArrowArray: IterRef {
-    type BaseArrayType: Array;
 
-    // Returns a typed iterator to the underlying elements of the array from an untyped Array reference.
-    fn iter_from_array_ref(b: &dyn Array) -> <Self as IterRef>::Iter<'_>;
+    #[inline]
+    #[doc(hidden)]
+    /// For internal use only
+    ///
+    /// TODO: this can be removed up by using arrow2::array::StructArray and
+    /// arrow2::array::UnionArray to perform the iteration for unions and structs
+    /// which should be possible if structs and unions are deserialized via scalars.
```

Comment:
@ncpenke Could you please expand on this? E.g. what needs to be done and what exactly does "deserialized via scalars" mean/imply?

Reply:
@aldanor The `arrow2::array::StructArray` iterator iterates over arrow2 scalars (https://github.com/jorgecarleitao/arrow2/blob/main/src/array/struct_/iterator.rs#L27). The concept of arrow scalars is described in the C++ API (https://arrow.apache.org/docs/cpp/api/scalar.html). I don't think they're an official part of the arrow spec, but an implementation choice to hold results of aggregate operations (for example sum of an array). 

However, they're also used to represent a single `Struct` element when iterating over a `StructArray`. I'm backtracking on this TODO though. I think existing approach is fine if the `unimplemented` into_iter hack is cleaned up.